### PR TITLE
feat: Add benchmark script to measure build speed with and w/o SAGP

### PR DESCRIPTION
**Pull request by romtsn**
_#skip-changelog_

### CODE REVIEW

File: scripts/benchmark/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Benchmark script
+
+This is a convenient script that benchmarks SAGP performance against the [duckduckgo-android](https://github.com/duckduckgo/Android) app.
+
+The script does the following:
+
+* Checks out the `duckduckgo` app locally
+* Runs [gradle-profiler](https://github.com/gradle/gradle-profiler) with predefined [scenarios](/duckduckgo/duckduckgo.scenarios)
```
Comment (romtsn) [inline scripts/benchmark/README.md:8]:
```suggestion
* Runs [gradle-profiler](https://github.com/gradle/gradle-profiler) with predefined [scenarios](/scripts/benchmark/duckduckgo/duckduckgo.scenarios)
```

File: scripts/benchmark/benchmark-build-speed [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+cd ..
+
+gradle-profiler --benchmark \
+ --project-dir Android \
+ --scenario-file duckduckgo/duckduckgo.scenarios \
+ --output-dir results/pre-sentry/
+
+git apply duckduckgo/add-sentry-to-duckduckgo.patch
+
+gradle-profiler --benchmark \
```
Comment (bruno-garcia) [inline scripts/benchmark/benchmark-build-speed:26]: Is it possible the second run relies on a "warm" gradle service that will skew the results? If so, does it make sense to have a "warm up phase" before the first run to get the service starting?

File: scripts/benchmark/benchmark-build-speed [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
```
Comment (bruno-garcia) [inline scripts/benchmark/benchmark-build-speed:1]: I suggest adding an extension ".sh" to be clear it's a shell script without the need to read it

File: scripts/benchmark/duckduckgo/add-sentry-to-duckduckgo.patch [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+     }
+@@ -28,6 +28,25 @@ allprojects {
+         google()
+         mavenCentral()
+     }
++
++}
++
++subprojects {
++     apply plugin: 'io.sentry.android.gradle'
```
Comment (bruno-garcia) [inline scripts/benchmark/duckduckgo/add-sentry-to-duckduckgo.patch:26]: Does it make sense to have the script bump the version? Or to use a locally installed package so we can run this in CI against a freshly built version? I guess the question should be: do you plan to run it in CI?

**Review by bruno-garcia — COMMENTED** [review]
Cool stuff!!

File: scripts/benchmark/benchmark-build-speed (hunk 2 above)
Comment (romtsn) [reply scripts/benchmark/benchmark-build-speed:26]: nope, if you take a look at [scenarios](https://github.com/getsentry/sentry-android-gradle-plugin/blob/feat/benchmark-script/scripts/benchmark/duckduckgo/duckduckgo.scenarios) they allow us to specify number of warm-ups and whether it should run with a warm or cold daemon. So this means every execution is being started from a clean state. However, I added `rm -rf gradle-user-home` just to make sure.

File: scripts/benchmark/duckduckgo/add-sentry-to-duckduckgo.patch (hunk 4 above)
Comment (romtsn) [reply scripts/benchmark/duckduckgo/add-sentry-to-duckduckgo.patch]: Solved it using includedBuild for the plugin (uses local version) and dynamic version for the main SDK.

As for running it on CI I will give it a shot, but don't wanna spend much time now

**Comment by romtsn** [comment]
@bruno-garcia wanna take another look at this? I added a CI workflow, which runs on PRs when something has changed in the `instrumentation` package, bcoz it takes ~2hrs to benchmark pre- and post- SAGP. The results are uploaded on gh actions, so we can manually download and check them for now.

File: scripts/benchmark/README.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+The script does the following:
+
+* Checks out the `duckduckgo` app locally
+* Runs [gradle-profiler](https://github.com/gradle/gradle-profiler) with predefined [scenarios](/scripts/benchmark/duckduckgo/duckduckgo.scenarios)
+* Applies git `.patch` which, in turn, applies sentry-android-gradle-plugin and sentry-android sdk to the duckduckgo app
+* Runs gradle-profiler with the same scenarios to measure the performance overhead of the Sentry gradle plugin
+* Saves the results under `/results/pre-sentry` and `/results/post-sentry`
+
+After launching the script, get some coffee and enjoy your time Slacking for the next 30 minutes or so :)
+Or you can cook some meat using your laptop as frying pan ;)
```
Comment (marandaneto) [inline scripts/benchmark/README.md:14]: What if I am a vegetarian?

File: scripts/benchmark/duckduckgo/add-sentry-to-duckduckgo.patch [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+     }
+@@ -37,8 +37,27 @@ allprojects {
+     }
+ }
+
++subprojects {
++    apply plugin: 'io.sentry.android.gradle'
++
++    sentry {
++        includeProguardMapping = true
```
Comment (marandaneto) [inline scripts/benchmark/duckduckgo/add-sentry-to-duckduckgo.patch:22]: That is enabled by default, better to remove it?

File: scripts/benchmark/duckduckgo/add-sentry-to-duckduckgo.patch [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
++        autoUploadProguardMapping = false
++
++        tracingInstrumentation {
++            enabled = true
++        }
```
Comment (marandaneto) [inline scripts/benchmark/duckduckgo/add-sentry-to-duckduckgo.patch:27]: That is enabled by default, better to remove it?

File: scripts/benchmark/README.md (hunk 5 above)
Comment (romtsn) [reply scripts/benchmark/README.md:14]: Then don't you dare running benchmarks on your machine :D

File: .github/workflows/benchmark-build-speed.yml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+          path: scripts/benchmark/Android
+          ref: 635cda1c1e5b7a6f8b5f04703b946663370d6c71
+          submodules: true
+
+      - name: Install SDKMAN, Gradle Profiler and Begin Profiling
+        run: |
+          curl -s "https://get.sdkman.io" | bash
+          source "$HOME/.sdkman/bin/sdkman-init.sh"
+          sdk install gradleprofiler
+          cd scripts/benchmark
```
[the remark this answers is not in the document]
Comment (romtsn) [reply .github/workflows/benchmark-build-speed.yml:29]: hm, but then I'd need to determine the latest version somehow, whereas sdkman does it for me. Do you know how to do that with `wget`?

[the remark this answers is not in the document]
Comment (cortinico) [reply .github/workflows/benchmark-build-speed.yml:29]: > but then I'd need to determine the latest version somehow, whereas sdkman does it for me. Do you know how to do that with `wget`?

If you wish to get the latest version, you'll have to get it from here: https://repo.gradle.org/ui/native/ext-releases-local/org/gradle/profiler/gradle-profiler/maven-metadata.xml which is potentially far from ideal.

However, I'll advise against using unpinned versions (e.g. with sdkman) as things can change as new version of `gradle-profiler` are released and they'll end up being unnoticed in the CI.

On the other hand, having a fixed version (either with a `wget` or with `sdk install gradleprofiler <version>`) will allow you to exclude this variability.

[the remark this answers is not in the document]
Comment (romtsn) [reply .github/workflows/benchmark-build-speed.yml]: Opted-in for still using sdkman but just pinned the version, thanks!

**Pull request merged**